chain/client/src/chunks_delay_tracker.rs

```diff
@@ -1,192 +1,129 @@
+                );
             }
         }
-        max_delay
-    }
-
-    fn update_chunks_metric(&mut self, head_height: BlockHeight) {
-        metrics::CHUNKS_RECEIVING_DELAY_US.set(self.get_max_delay(head_height).as_micros() as i64);
     }
 
-    // Computes the difference between the latest block we are aware of and the current head.
-    fn get_blocks_ahead(&mut self, head_height: BlockHeight) -> u64 {
-        if let Some((&latest, _)) = self.heights.iter().next_back() {
-            latest.saturating_sub(head_height)
-        } else {
-            0
+    pub fn requested_chunk(&mut self, chunk_hash: &ChunkHash, timestamp: Instant) {
+        match self.chunks_in_progress.get_mut(chunk_hash) {
+            Some(entry) => {
+                entry.chunk_requested.get_or_insert(timestamp);
+            }
+            None => {
+                self.chunks_in_progress.put(
+                    chunk_hash.clone(),
+                    ChunkInProgress { chunk_received: None, chunk_requested: Some(timestamp) },
+                );
+            }
         }
     }
 
-    fn update_blocks_ahead_metric(&mut self, head_height: BlockHeight) {
-        metrics::BLOCKS_AHEAD_OF_HEAD.set(self.get_blocks_ahead(head_height) as i64);
+    pub fn finish_block_processing(
+        &mut self,
+        processed_block_hash: &CryptoHash,
+        chunks: &[(ChunkHash, ShardId)],
+    ) {
+        self.update_block_chunks_metric(processed_block_hash, &chunks);
+        self.update_chunks_metric(&chunks);
+        self.blocks_received.remove(&processed_block_hash);
     }
 
-    fn update_metrics(&mut self, head_height: BlockHeight) {
-        self.update_chunks_metric(head_height);
-        self.update_blocks_ahead_metric(head_height);
+    pub fn get_chunks(block: &Block) -> Vec<(ChunkHash, ShardId)> {
+        block.chunks().iter().map(|chunk| (chunk.chunk_hash(), chunk.shard_id())).collect()
     }
 
-    pub fn add_block_timestamp(
+    fn update_block_chunks_metric(
         &mut self,
-        height: BlockHeight,
-        head_height: BlockHeight,
-        timestamp: Instant,
+        block_hash: &CryptoHash,
+        chunks: &[(ChunkHash, ShardId)],
     ) {
-        self.remove_old_entries(head_height);
-        if height >= head_height {
-            self.heights.entry(height).or_default().block_received.get_or_insert(timestamp);
+        if let Some(block_received) = self.blocks_received.get(&block_hash) {
+            for (chunk_hash, shard_id) in chunks {
+                let chunk_in_progress = self.chunks_in_progress.peek(&chunk_hash.clone());
+                if let Some(chunk_in_progress) = chunk_in_progress {
+                    if let Some(chunk_received) = chunk_in_progress.chunk_received {
+                        metrics::BLOCK_CHUNKS_DELAY
+                            .with_label_values(&[&format!("{}", shard_id)])
+                            .observe(
+                                chunk_received
```

Comment:
It's probably better to track the time between "block_received" and "chunk_requested" instead of between "block_received" and "chunk_received", since we have another metric for between "chunk_requested" and "chunk_received"

Reply:
Done